core/src/main/resources/jenkins/formelementpath/form-element-path.js

```diff
@@ -0,0 +1,186 @@
+                    break;
+                case "file":
+                    // to support structured form submission with file uploads,
+                    // rename form field names to unique ones, and leave this name mapping information
+                    // in JSON. this behavior is backward incompatible, so only do
+                    // this when
+                    p = findParent(e);
+                    if (e.getAttribute("jsonAware") != null) {
+                        var on = e.getAttribute("originalName");
+                        if (on != null) {
+                            addProperty(p, on, e);
+                        } else {
+                            addProperty(p, e.name, e);
+                        }
+                    }
+                    break;
+                // otherwise fall through
+                default:
+                    p = findParent(e);
+                    addProperty(p, e.name, e);
+                    break;
+            }
+        }
+
+        function annotate(e, path) {
+            e.setAttribute("path", path);
+            var o = e.formDom || {};
+            for (var key in o) {
+                var v = o[key];
+
+                function child(v, i) {
+                    var suffix = null;
+                    var newKey = key;
+                    if (v.parentNode.className && v.parentNode.className.indexOf("one-each") > -1 && v.parentNode.className.indexOf("honor-order") > -1) {
+                        suffix = v.getAttribute("descriptorId").split(".").pop()
+                    } else if (v.getAttribute("type") == "radio") {
+                        suffix = v.value
+                        while (newKey.substring(0, 8) == 'removeme')
+                            newKey = newKey.substring(newKey.indexOf('_', 8) + 1);
+                    } else if (v.getAttribute("suffix") != null) {
+                        suffix = v.getAttribute("suffix")
+                    } else {
+                        if (i > 0)
+                            suffix = i;
+                    }
+                    if (suffix == null) suffix = "";
+                    else suffix = '[' + suffix + ']';
+
+                    annotate(v, path + "/" + newKey + suffix);
+                }
+
+                if (v instanceof Array) {
+                    var i = 0;
+                    v.forEach(function (v) {
+                        child(v, i++)
+                    })
+                } else {
+                    child(v, 0)
+                }
+            }
+
+        }
+
+        annotate(form, "");
+
+        // clean up
+        for (i = 0; i < doms.length; i++)
+            doms[i].formDom = null;
+
+        return true;
+    }
+
+    function applyAll() {
+        document.querySelectorAll("FORM").forEach(function (e) {
+            buildFormTree(e);
+        })
+    }
+
+    applyAll();
+
+    // run this periodically to cope with DOM changes
+    window.setInterval(applyAll, 1000);
```

Comment:
this is what I'm most worried about having it on by default, every second recalculating the paths.

I tried removing it but its needed for repeatable elements to have the index added and re-calculated.

Reply:
Jenkins already has runtime performance issues and periodic queries, I would avoid this as much as possible. 

I'd suggest maybe registering this as a layout update callback (https://github.com/jenkinsci/jenkins/blob/3e8afc53de7824a8d6fc113dff655aeb7174c55f/war/src/main/webapp/scripts/hudson-behavior.js#L2487). Most widgets already use it after adding new stuff (https://github.com/jenkinsci/jenkins/search?q=layoutUpdateCallback). Only `repeatable` is missing I think.

Reply:
it's already registered a couple of lines down, I'll take a look at adding it to repeatable to see if it removes the need for this.

Reply:
It should probably fix it for most plugins. Don't know if it's a problem that rogue plugins with custom widgets do not have this tbh

Reply:
> Jenkins already has runtime performance issues and periodic queries, I would avoid this as much as possible.

As in https://github.com/jenkinsci/jenkins/pull/5926#pullrequestreview-809223262 you should _either_

* Add the new HTML attributes or whatever unconditionally during Jelly generation, deleting all the JavaScript hacks.
* Or disable all of this by default in production code.

Reply:
Will probably disable by default but also look to add repeatable to layout so the timer bit can be removed.

The timer bit made debugging this code a pain.

Reply:
Maybe start with the current code copied-and-pasted from the plugin but disabled by default. That at least solves the immediate issue, we can gut the plugin, and simplify ATH to not try to load it.

Then as a follow-up look into whether the actual behavior can be simplified by inlining into `lib/form/*.jelly` and/or `hudson-behavior.js`.